File: 0056-merge-intervals/0056-merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        if (intervals.length <= 1) {
            return intervals;
        }

        // Sort the intervals by their start time
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));

        // Initialize result array with same size as input (we'll truncate later)
        List<int[]> res = new ArrayList<>();
        res.add(intervals[0]); // Add first interval




        for (int i = 1; i < intervals.length; i++) {
            int[] last = res.get(res.size() - 1);
            int[] current = intervals[i];

            if (last[1] >= current[0]) {
                last[1] = Math.max(last[1], current[1]);
            } else {
                res.add(current);
            }
        }
     return res.toArray(new int[res.size()][]);   
      }
}

File: 0056-merge-intervals/0056-merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key = lambda x: x[0])
        newList = [intervals[0]]

        for i in range(1, len(intervals)):
            if intervals[i][0] <= newList[-1][1]:
                newList[-1][0] = min(intervals[i][0], newList[-1][0])
                newList[-1][1] = max(intervals[i][1], newList[-1][1])
            else:
                newList.append(intervals[i])

        return newList 
        